pydantic/main.py

```diff
@@ -56,16 +56,35 @@
 _base_class_defined = False
 
 
+def build_generic_metadata(
```

Comment:
Probably want to make this private at least, if not move into the `_generics` module. Since it might be convenient to see this logic while working on `ModelMetaclass` I wasn't sure if it was a good idea to pull it out of this module.

Reply:
If this is a new function, any reason why it shouldn't be private? It seems to be used only in this module.

I like `update_generic_metadata_for_class`. I am curious as to why it's updating the metadata instead of "collecting it. I generally prefer to collect it and let the caller update it via `metadata = {**metadata, **collect_metadata_from_cls(cls)}` or something like that.

Reply:
There are two "scenarios" where this `build_generic_metadata`/`update_generic_metadata_for_class` needs to do non-trivial work:

1. When you create a "new" generic class (i.e., when you create a new class that subclasses `BaseModel` _and_ `typing.Generic`). In this case, the `__pydantic_generic_metadata__` argument to `ModelMetaclass.__new__` will be `None`.
2. When you parametrize a generic class (i.e., call `BaseModel.__class_getitem__`), which in turn calls `create_generic_submodel`, this calls `ModelMetaclass.__new__` with a non-trivial `__pydantic_generic_metadata__`.

This implementation allows much/most of the logic to be shared between the two paths. I am not yet sure if this can be simplified; my intuition is it will require an annoying amount of logic duplication, but I'm looking into it.